### PR TITLE
enh(javascript) highlight class methods

### LINKED ISSUE #1987: (javascript) Member functions are not highlighted in ES6 classes
This is how I use highlight.js in React (combined with marked), but I find that the functions are not highlighted:
```
marked.setOptions({
  highlight: (code) => hljs.highlightAuto(code).value
});


class Detail extends Component {
  constructor(props) {
    super(props);
    this.state = {
      detail: ''
    }
  }

 
  render () {
    const markdown = marked(this.state.detail);
    return (
      <Fragment>
        <Header></Header>
        <DetailWrapper 
        dangerouslySetInnerHTML = {{__html: markdown}} 
        />
        <ScrollToTop/>
      </Fragment>
    )
  }
}
```
and this is how it looks like after rendered(functions are not highlighted ):
image

### PR DESCRIPTION
**Pull request by joshgoebel**
Resolved #1987.  Never say never.

Doesn't deal with the edge case that TECHNICALLY `blah() {}` could be a function call FOLLOWED by a block as this is a very unusual coding style.

- Also changes "constructor" to be a function/title rather than a keyword

### CODE REVIEW

File: src/languages/javascript.js [hunk 1]
```diff
@@ -325,6 +325,23 @@ export default function(hljs) {
         ],
         illegal: /%/
       },
+      {
+        className: 'function',
+        // we have to count the parens to make sure we actually have the
+        // correct bounding ( ) before the =>.  There could be any number of
```
Comment (egor-rogov) [inline src/languages/javascript.js:331]: Should be `before the {` instead of `=>` here.

File: src/languages/javascript.js [hunk 2]
```diff
@@ -325,6 +325,23 @@ export default function(hljs) {
...
+        // sub-expressions inside also surrounded by parens.
+        begin: hljs.UNDERSCORE_IDENT_RE + '(\\([^(]*' +
+              '(\\([^(]*' +
+                '(\\([^(]*' +
+                '\\))?' +
+              '\\))?' +
+            '\\))\\s*{',
```
Comment (egor-rogov) [inline src/languages/javascript.js:338]: Can you give me an example of nested parens here?

Comment (joshgoebel) [reply src/languages/javascript.js:338]: Huh?  ((3))

Comment (egor-rogov) [reply src/languages/javascript.js:338]: And what is the meaning of it as a function parameter?
Okay, I really just want to say that the regex doesn't handle `((3) )`, which I believe to be correct.

Comment (joshgoebel) [reply src/languages/javascript.js:338]: Example: 

```
  other(a = (3+2)) {
    console.log(a)
  }
```

I suppose we can eat spaces on both sides, though that's unusual. :-)

**Comment by joshgoebel** [comment]
@egor-rogov Ping.

File: src/languages/javascript.js [hunk 3]
```diff
@@ -250,8 +250,8 @@ export default function(hljs) {
             begin: '(\\([^(]*' +
               '(\\([^(]*' +
                 '(\\([^(]*' +
-                '\\))?' +
-              '\\))?' +
+                '\\)\\s*)?' +
+              '\\)\\s*)?' +
```
Comment (egor-rogov) [inline src/languages/javascript.js:254]: Not only spaces, I guess? Consider this:
```js
class Test {
  test(a = (3+2), b = 1) {
    console.log(a + b)
  }
}
```
Looks like `[^)]*` will do the job.

**Review by egor-rogov — CHANGES_REQUESTED** [review]

File: src/languages/javascript.js (hunk 3 above)
Comment (joshgoebel) [reply src/languages/javascript.js]: Actually you also have to account for later expressions surrounded by ()... the new pattern tries to match:

```
(
.*(  
.*(
.*  
).*
).*
)
```

Where . means not `()`.

**Review by egor-rogov — APPROVED** [review]
This feels right :+1:

**Pull request merged**